### DISCUSSION TITLE
Pre-commit integration

### DISCUSSION THREAD

**Discussion by Wyko** [callowayproject/bump-my-version#257] [Ideas]
# Feature request

I noticed that in issue #62 there was mention of pre-commit hooks, and it got me thinking: Would you be able to create a hook for [pre-commit](https://pre-commit.com/) that automatically bumps version on each commit? 

From personal experience, pre-commit (the package) doesn't play nicely with manual pre-commit hooks, so it would be nice if we could integrate everything under their ecosystem. It would definitely help my workflow a lot!

**Comment by coordt** [comment]
I can't say that I've ever heard of a workflow that increments the version _on commit._ With the current workload I don't see this becoming a priority unless others are also interested.

**Comment by azataiot** [comment]
Incrementing the version on commit will be too much overkill. 

Normally a commit is a simple state change that recorded.But a version is a complete state change in the code, like a feature, a new endpoint or a service etc.

**Comment by lnxpy** [comment]
I'm just imagining how would your Linux kernel version look like then. It has over 1.2M commits by now. ([torvalds/linux](https://github.com/torvalds/linux))

Technically, that's not a good idea. Neither is an awful one. In some very certain cases and situations, it might be helpful. As a solution, you are able to run shell scripts in your `.pre-commit-config.yaml` file.

You can simply put `bump-my-version major/minor/patch` in there, but be **VERY** careful as you might get trapped in a circular infinite loop of pre-commit checking and version-bumping.

When you `git commit`, your hook will get triggered. If the commit flag is enabled for `bump-my-version`, it'll commit once again, which will cause your hook to execute right after, bump the version and so on and so forth!
